## Introduction

In this project we will create data visualization for Baseball Data set using Tableau and we will analyze various variables such as batting average, handedness, player’s height,weight and homeruns ,mutual relationship of these variables and how they determine player’s perfromance.

### Analyze

So, first let's check the basic structure of the **Baseball Data** Dataset using pandas

In [53]:
import pandas as pd

In [54]:
# read csv as a pandas dataframe
df = pd.read_csv('baseball_data.csv')

In [55]:
#Checking total rows and columns
df.shape

(1157, 6)

In [56]:
#Checking the Column names
df.head(2)

name handedness  height  weight   avg  HR
0       Tom Brown          R      73     170  0.00   0
1  Denny Lemaster          R      73     182  0.13   4

So there are 1157 rows and 6 Columns in this dataset.

- We will check for duplicate names.
- We will fix Column names & their values as they are not clear and contains abreviations like HR, avg,L,R,B etc.


Let's check if any duplicate names are present and their counts :

In [57]:
#Checking Duplicate names count
df.name.duplicated().sum()

6

There are 6 duplicate names. Now we need to check the duplicate name row details

In [58]:
#Checking rows with duplicate names
df[df['name'].duplicated()]

name handedness  height  weight    avg  HR
245     Dave Roberts          L      75     195  0.194   7
454   Bobby Mitchell          R      75     185  0.235  21
645       Mike Brown          R      74     195  0.000   0
704   Dave Stapleton          R      73     178  0.271  41
714       Jim Wright          R      73     165  0.000   0
939  Mel Stottlemyre          R      72     190  0.000   0

### Task1 : Fix the duplicate names

### Clean

#### Define

We will modify the above found rows to fix duplicate issue by adding 2 on the end of duplicate names.

#### Code

In [59]:
#Modifying the player names 
df.loc[df.duplicated(subset=['name'], keep="first"), 'name'] = \
df.loc[df.duplicated(subset=['name'], keep="first"), 'name'].apply(lambda x: x+'2')

#### Test

Let's check if our change worked fine. We will check again if there appears any duplicate name values.

In [60]:
#Checking if any duplicate names
df[df['name'].duplicated()]

Empty DataFrame
Columns: [name, handedness, height, weight, avg, HR]
Index: []

Good! We have no duplicate row appeared as seen above. Lets verify the names after the modification

In [61]:
#Checking the changes made to duplicate names

#Create list of duplicate names
Name = ['Dave Roberts','Dave Roberts2','Bobby Mitchell','Bobby Mitchell2',
       'Mike Brown','Mike Brown2','Dave Stapleton','Dave Stapleton',
       'Jim Wright','Jim Wright2','Mel Stottlemyre','Mel Stottlemyre2']

#Checking row details for all above names
df.loc[df.name.isin(Name)]

name handedness  height  weight    avg  HR
206      Dave Roberts          R      75     215  0.239  49
245     Dave Roberts2          L      75     195  0.194   7
375   Mel Stottlemyre          R      73     178  0.160   7
453    Bobby Mitchell          L      70     170  0.243   3
454   Bobby Mitchell2          R      75     185  0.235  21
644        Mike Brown          R      74     195  0.265  23
645       Mike Brown2          R      74     195  0.000   0
703    Dave Stapleton          L      73     185  0.000   0
713        Jim Wright          R      77     205  0.000   0
714       Jim Wright2          R      73     165  0.000   0
939  Mel Stottlemyre2          R      72     190  0.000   0

We can see that one of the name has been modified. Let's save the modified data as different dataframe

In [62]:
#Saving the modified dataframe
df1 = df.to_csv('baseball_data_clean.csv', index=False)

In [63]:
# read csv as a pandas new dataframe
df1 = pd.read_csv('baseball_data_clean.csv')
df1

name handedness  height  weight    avg   HR
0             Tom Brown          R      73     170  0.000    0
1        Denny Lemaster          R      73     182  0.130    4
2             Joe Nolan          L      71     175  0.263   27
3           Denny Doyle          L      69     175  0.250   16
4         Jose Cardenal          R      70     150  0.275  138
5             Mike Ryan          R      74     205  0.193   28
6        Fritz Peterson          B      72     185  0.159    2
7          Dick Bertell          R      72     200  0.250   10
8            Rod Kanehl          R      73     180  0.241    6
9          Ozzie Osborn          R      74     195  0.000    0
10          John LaRose          L      73     185  0.000    0
11            Jeff Byrd          R      75     195  0.000    0
12         Juan Bonilla          R      69     170  0.256    7
13       Frank Tepedino          L      71     185  0.241    6
14        Jerry Kutzler          L      73     175  0.000    0
15            Rick Wise          R      73     180  0.195   15
16        Darrell Evans          L      74     200  0.248  414
17        Buck Martinez          R      70     190  0.225   58
18      John Wockenfuss          R      72     190  0.262   86
19         Tommy McCraw          L      72     183  0.246   75
20         Jeff Barkley          B      75     185  0.000    0
21         Eric Wilkins          R      73     190  0.000    0
22          Britt Burns          L      77     215  0.000    0
23             Lee Lacy          R      73     175  0.286   91
24            Pete Rose          B      71     192  0.303  160
25         Danny Walton          R      72     195  0.223   28
26            Bobby Cox          R      71     180  0.225    9
27         Brad Wellman          R      72     170  0.231    6
28          Jeff Newman          R      74     215  0.224   63
29       Pedro Gonzalez          R      72     176  0.244    8
...                 ...        ...     ...     ...    ...  ...
1127       Dave Johnson          R      73     183  0.000    0
1128       Gail Hopkins          L      70     198  0.266   25
1129        Doug Frobel          L      76     196  0.201   20
1130      Luis Melendez          R      72     165  0.248    9
1131    Jerry Augustine          L      72     183  0.000    0
1132       Bill Plummer          R      73     190  0.188   14
1133       Buck Rodgers          B      74     190  0.232   31
1134        Phil Niekro          R      73     180  0.169    7
1135         Jeff Holly          L      77     210  0.000    0
1136       Tom Poquette          L      70     175  0.268   10
1137        Eric Hetzel          R      75     175  0.000    0
1138   Brian Asselstine          L      73     175  0.254   12
1139        Rick Miller          L      72     175  0.269   28
1140         Ed Charles          R      70     170  0.263   86
1141         Luis Gomez          R      69     150  0.210    0
1142     Dave Lemanczyk          R      76     235  0.000    0
1143           Al Jones          R      76     210  0.000    0
1144      Floyd Rayford          R      70     190  0.244   38
1145          Sam Hinds          R      78     215  0.000    0
1146    Tony Conigliaro          R      75     185  0.264  166
1147  Clay Christiansen          R      77     205  0.000    0
1148           Jim Barr          R      75     205  0.162    1
1149         Jerry Remy          L      69     165  0.275    7
1150   Larry Rothschild          L      74     180  0.000    0
1151        Frank Duffy          R      73     180  0.232   26
1152         Gary Lance          B      75     195  0.000    0
1153       Tom O'Malley          L      72     180  0.256   13
1154        Phil Garner          R      70     175  0.260  109
1155     Larry Pashnick          R      75     205  0.000    0
1156       Todd Fischer          R      70     170  0.000    0

[1157 rows x 6 columns]

Now we have saved modified data into different data frame 'baseball_data_clean.csv' but there are still few more modifications needed in our new dataset. 

### Task2 : Fix Column names & abbreviated values 

### Clean

#### Define

To fix the column names we will map them to the desired ones. Let's make the column names more informative.

#### Code

In [64]:
#Changing the column names to make their meaning more clear
df1 = df1.rename(columns={'height': 'Height in inches', 'weight': 'Weight in pounds',
                         'name': 'Player Name','avg' : 'Batting Average%','HR':'HomeRuns','handedness':'Handedness'})

In [65]:
# replacing the abbreviations with proper names
df1.Handedness.replace(['L', 'R','B'], ['Left Handed','Right Handed','Both'], inplace=True)

In [66]:
#Rounding up average upto 3 decimal places
decimals = 3    
df1['Batting Average%'] = df1['Batting Average%'].apply(lambda x: round(x, decimals))

In [67]:
#Saving again all these changes
df1.to_csv('baseball_data_clean.csv', index=False)

#### Test

In [68]:
#Checking column names
df1.head(5)

Player Name    Handedness  Height in inches  Weight in pounds  \
0       Tom Brown  Right Handed                73               170   
1  Denny Lemaster  Right Handed                73               182   
2       Joe Nolan   Left Handed                71               175   
3     Denny Doyle   Left Handed                69               175   
4   Jose Cardenal  Right Handed                70               150   

   Batting Average%  HomeRuns  
0             0.000         0  
1             0.130         4  
2             0.263        27  
3             0.250        16  
4             0.275       138

As we see above, the columns names looks more clearer and average is rounded up to 2 decimal places. 
Now we are ready to create visualizations using cleaned dataset.

In [69]:
#Final check of the dataframe
df1

Player Name    Handedness  Height in inches  Weight in pounds  \
0             Tom Brown  Right Handed                73               170   
1        Denny Lemaster  Right Handed                73               182   
2             Joe Nolan   Left Handed                71               175   
3           Denny Doyle   Left Handed                69               175   
4         Jose Cardenal  Right Handed                70               150   
5             Mike Ryan  Right Handed                74               205   
6        Fritz Peterson          Both                72               185   
7          Dick Bertell  Right Handed                72               200   
8            Rod Kanehl  Right Handed                73               180   
9          Ozzie Osborn  Right Handed                74               195   
10          John LaRose   Left Handed                73               185   
11            Jeff Byrd  Right Handed                75               195   
12         Juan Bonilla  Right Handed                69               170   
13       Frank Tepedino   Left Handed                71               185   
14        Jerry Kutzler   Left Handed                73               175   
15            Rick Wise  Right Handed                73               180   
16        Darrell Evans   Left Handed                74               200   
17        Buck Martinez  Right Handed                70               190   
18      John Wockenfuss  Right Handed                72               190   
19         Tommy McCraw   Left Handed                72               183   
20         Jeff Barkley          Both                75               185   
21         Eric Wilkins  Right Handed                73               190   
22          Britt Burns   Left Handed                77               215   
23             Lee Lacy  Right Handed                73               175   
24            Pete Rose          Both                71               192   
25         Danny Walton  Right Handed                72               195   
26            Bobby Cox  Right Handed                71               180   
27         Brad Wellman  Right Handed                72               170   
28          Jeff Newman  Right Handed                74               215   
29       Pedro Gonzalez  Right Handed                72               176   
...                 ...           ...               ...               ...   
1127       Dave Johnson  Right Handed                73               183   
1128       Gail Hopkins   Left Handed                70               198   
1129        Doug Frobel   Left Handed                76               196   
1130      Luis Melendez  Right Handed                72               165   
1131    Jerry Augustine   Left Handed                72               183   
1132       Bill Plummer  Right Handed                73               190   
1133       Buck Rodgers          Both                74               190   
1134        Phil Niekro  Right Handed                73               180   
1135         Jeff Holly   Left Handed                77               210   
1136       Tom Poquette   Left Handed                70               175   
1137        Eric Hetzel  Right Handed                75               175   
1138   Brian Asselstine   Left Handed                73               175   
1139        Rick Miller   Left Handed                72               175   
1140         Ed Charles  Right Handed                70               170   
1141         Luis Gomez  Right Handed                69               150   
1142     Dave Lemanczyk  Right Handed                76               235   
1143           Al Jones  Right Handed                76               210   
1144      Floyd Rayford  Right Handed                70               190   
1145          Sam Hinds  Right Handed                78               215   
1146    Tony Conigliaro  Right Handed                75               185   
1147  Cl